src/Features/Core/Portable/Completion/Providers/AbstractTypeImportCompletionProvider.cs

```diff
@@ -0,0 +1,316 @@
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Completion.Providers
+{
+    internal abstract partial class AbstractTypeImportCompletionProvider : CommonCompletionProvider
+    {
+        private readonly SyntaxAnnotation _annotation = new SyntaxAnnotation();
+
+        protected abstract Task<SyntaxContext> CreateContextAsync(Document document, int position, CancellationToken cancellationToken);
+        protected abstract ImmutableHashSet<string> GetNamespacesInScope(SyntaxNode location, SemanticModel semanticModel, CancellationToken cancellationToken);
+
+        private readonly ITypeImportCompletionService _typeImportCompletionService;
+        private readonly IExperimentationService _experimentationService;
+
+        public AbstractTypeImportCompletionProvider(Workspace workspace)
+            : base()
+        {
+            _typeImportCompletionService = workspace.Services.GetService<ITypeImportCompletionService>();
+            _experimentationService = workspace.Services.GetService<IExperimentationService>();
+        }
+
+        public override async Task ProvideCompletionsAsync(CompletionContext completionContext)
+        {
+#if DEBUG
+            DebugObject.DebugClear();
+            var tick = System.Environment.TickCount;
+#endif
+            var document = completionContext.Document;
+            var position = completionContext.Position;
+            var cancellationToken = completionContext.CancellationToken;
+
+            var importCompletionOptionValue = completionContext.Options.GetOption(CompletionOptions.ShowImportCompletionItems, document.Project.Language);
+
+            // Don't trigger import completion if the option value is "default" and the experiment is disabled for the user. 
+            if (importCompletionOptionValue == false ||
+                (importCompletionOptionValue == null && _experimentationService?.IsExperimentEnabled(WellKnownExperimentNames.TypeImportCompletion) != true))
+            {
+                return;
+            }
+
+            var syntaxContext = await CreateContextAsync(document, position, cancellationToken).ConfigureAwait(false);
+
+            var items = await GetCompletionItemsAsync(document, syntaxContext, position, cancellationToken).ConfigureAwait(false);
+            completionContext.AddItems(items);
+#if DEBUG
+            DebugObject.debug_total_time_with_ItemCreation = System.Environment.TickCount - tick;
+#endif
+            return;
+        }
+
+        private async Task<ImmutableArray<CompletionItem>> GetCompletionItemsAsync(Document document, SyntaxContext context, int position, CancellationToken cancellationToken)
+        {
+            if (context.IsTypeContext)
+            {
+                using (Logger.LogBlock(FunctionId.Completion_TypeImportCompletionProvider_GetCompletionItemsAsync, cancellationToken))
+                {
+                    var project = document.Project;
+                    var node = context.LeftToken.Parent;
```

Comment:
so, LeftToken might not exist.  so `.Parent` would be null.  is the later code ok with this.

Reply:
Fixed